common/src/main/java/org/conscrypt/ClientSessionContext.java

```diff
@@ -131,6 +147,42 @@ private NativeSslSession getSession(String host, int port) {
         return null;
     }
 
+    private void putSession(HostAndPort key, NativeSslSession session) {
+        synchronized (sessionsByHostAndPort) {
+            List<NativeSslSession> sessions = sessionsByHostAndPort.get(key);
+            if (sessions == null) {
+                sessions = new ArrayList<NativeSslSession>();
+                sessionsByHostAndPort.put(key, sessions);
+            }
+            // To maintain the invariant that single- and multi-use sessions aren't
+            // mixed, check what the current list contains and remove those sessions if
+            // they're of the other type.
+            if (sessions.size() > 0) {
+                if (sessions.get(0).isSingleUse() != session.isSingleUse()) {
+                    while (!sessions.isEmpty()) {
+                        removeSession(sessions.get(0));
```

Comment:
Does this actually work? Did you mean to call the version with a key parameter? This one keys by session ID.

Edit: Okay, I think now I understand how this logic works. This session cache implementation probably needs to be rewritten before you can do TLS 1.3. It looks like `AbstractSessionContext` keys by session ID always and then it exposes hooks for the client session cache to maintain a parallel host-keyed one.

That's not how client session caches work. A client  should not maintain any key by session ID. The client cannot even assume that sessions have unique session IDs because servers can maliciously pick those values. Only the server gets to assume that because it generates the IDs.

Moreover, there aren't any session IDs in TLS 1.3 and BoringSSL does not even fill in the field. I see you instead have checks for the ID being empty, but that just means you aren't going to cache anything, no?

All this also has implications for relying on `AbstractSessionContext` to enforce size limits for you.

Reply:
For missing session IDs, I've got code in my TLS 1.3 support branch (which could be moved into this change) that hashes the session ticket and uses that for ID purposes if the session doesn't have an ID, much like BoringSSL does internally for TLS 1.2 sessions.  So that should enable TLS 1.3 sessions to be cached.

In terms of the ID-keyed client-side cache, it probably should be rewritten, but I'd rather do that in a separate change, since I think it's orthogonal to this.  The only real reason we have the ID-keyed cache is because there's a standard Java API that exposes it ([`SSLSessionContext`](https://docs.oracle.com/javase/7/docs/api/javax/net/ssl/SSLSessionContext.html)), so it requires some thought on what we want to do with it.

I have updated this class so that if a second session comes in with the same ID it removes the first, so that multiple sessions with the same ID don't result in the per-host/port cache getting out of sync with the per-ID cache.